pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapFillGroupByDataTableReducer.java

```diff
@@ -0,0 +1,490 @@
+
+  private final QueryContext _queryContext;
+  private final AggregationFunction[] _aggregationFunctions;
+  private final int _numAggregationFunctions;
+  private final List<ExpressionContext> _groupByExpressions;
+  private final int _numGroupByExpressions;
+  private final int _numColumns;
+  private final DateTimeGranularitySpec _dateTimeGranularity;
+  private final DateTimeFormatSpec _dateTimeFormatter;
+  private final long _startMs;
+  private final long _endMs;
+  private final Set<Key> _groupByKeys;
+  private final Map<Key, Object[]> _previousByGroupKey;
+  private final int _numOfGroupByKeys;
+  private final List<Integer> _groupByKeyIndexes;
+  private final boolean [] _isGroupBySelections;
+  private int _timeBucketIndex = -1;
+
+  GapFillGroupByDataTableReducer(QueryContext queryContext) {
+    Preconditions.checkArgument(
+        queryContext.getBrokerRequest().getPinotQuery() != null, "GapFill can only be applied to sql query");
+    _queryContext = queryContext;
+    _aggregationFunctions = queryContext.getAggregationFunctions();
+    assert _aggregationFunctions != null;
+    _numAggregationFunctions = _aggregationFunctions.length;
+    _groupByExpressions = queryContext.getGroupByExpressions();
+    assert _groupByExpressions != null;
+    _numGroupByExpressions = _groupByExpressions.size();
+    _numColumns = _numAggregationFunctions + _numGroupByExpressions;
+
+    ExpressionContext gapFillSelection = null;
+    for (ExpressionContext expressionContext : _queryContext.getSelectExpressions()) {
+      if (GapfillUtils.isPostAggregateGapfill(expressionContext)) {
+        gapFillSelection = expressionContext;
+        break;
+      }
+    }
+
+    List<ExpressionContext> args = gapFillSelection.getFunction().getArguments();
+    Preconditions.checkArgument(
+        args.size() == 5, "PostAggregateGapFill does not correct number of arguments.");
+    Preconditions.checkArgument(
+        args.get(1).getLiteral() != null, "The second argument of PostAggregateGapFill should be TimeFormatter.");
+    Preconditions.checkArgument(
+        args.get(2).getLiteral() != null, "The third argument of PostAggregateGapFill should be start time.");
+    Preconditions.checkArgument(
+        args.get(3).getLiteral() != null, "The fourth argument of PostAggregateGapFill should be end time.");
+    Preconditions.checkArgument(
+        args.get(4).getLiteral() != null, "The fifth argument of PostAggregateGapFill should be time bucket size.");
+
+    boolean orderByTimeBucket = false;
+    if (_queryContext.getOrderByExpressions() != null) {
+      for (OrderByExpressionContext expressionContext : _queryContext.getOrderByExpressions()) {
+        if (expressionContext.getExpression().equals(gapFillSelection)) {
+          orderByTimeBucket = true;
+          break;
+        }
+      }
+    }
+
+    Preconditions.checkArgument(
+        orderByTimeBucket, "PostAggregateGapFill does not work if the time bucket is not ordered.");
```

Comment:
Should we enforce , it should only be ordered by time bucket ?

Reply:
It might break if its ordered by more than one field along with timebucket?

Reply:
The result should be ordered by time bucket, but it can be ordered by more than one field along with time bucket.

Reply:
Could you add a test for this ?